producer.go

```diff
@@ -150,12 +155,47 @@ func (p *producer) Run(fetchCtx, workCtx context.Context, statusFunc producerSta
 	// TODO: fetcher should have some jitter in it to avoid stampeding issues.
 	fetchLimiter := chanutil.NewDebouncedChan(fetchCtx, p.config.FetchCooldown)
 
+	handleJobControlNotification := func(topic notifier.NotificationTopic, payload string) {
+		var decoded jobControlPayload
+		if err := json.Unmarshal([]byte(payload), &decoded); err != nil {
+			p.Logger.ErrorContext(workCtx, p.Name+": Failed to unmarshal job control notification payload", slog.String("err", err.Error()))
+			return
+		}
+		if string(decoded.Action) == string(jobControlActionCancel) && decoded.Queue == p.config.QueueName && decoded.JobID > 0 {
+			select {
+			case p.cancelCh <- decoded.JobID:
+			default:
+				p.Logger.WarnContext(workCtx, p.Name+": Job cancel notification dropped due to full buffer", slog.Int64("job_id", decoded.JobID))
+			}
+			return
+		}
+		p.Logger.DebugContext(workCtx, p.Name+": Received job control notification with unknown action or other queue",
+			slog.String("action", string(decoded.Action)),
+			slog.Int64("job_id", decoded.JobID),
+			slog.String("queue", decoded.Queue),
+		)
+	}
+	sub := p.config.Notifier.Listen(notifier.NotificationTopicJobControl, handleJobControlNotification)
+	defer sub.Unlisten()
+
 	p.fetchAndRunLoop(fetchCtx, workCtx, fetchLimiter, statusFunc)
 	statusFunc(p.config.QueueName, componentstatus.ShuttingDown)
 	p.executorShutdownLoop()
 	statusFunc(p.config.QueueName, componentstatus.Stopped)
 }
 
+type jobControlAction string
+
+const (
+	jobControlActionCancel jobControlAction = "cancel"
```

Comment:
OOC, are there any other actions you were expecting to come in?

Reply:
I don't have anything else in mind that's specific to a single in-progress job. You could retry a single job immediately (no matter its state), but that doesn't require a notification because it's just a matter of updating the job row.